core/src/main/java/org/elasticsearch/index/mapper/Uid.java

```diff
@@ -123,4 +126,169 @@ public static String createUid(String type, String id) {
         return type + DELIMITER + id;
     }
 
+    private static final int UTF8 = 0xff;
+    private static final int NUMERIC = 0xfe;
+    private static final int BASE64_ESCAPE = 0xfd;
+
+    static boolean isURLBase64WithoutPadding(String id) {
+        // We are not lenient about padding chars ('=') otherwise
+        // 'xxx=' and 'xxx' could be considered the same id
+        final int length = id.length();
+        switch (length & 0x03) {
+        case 0:
+            break;
+        case 1:
+            return false;
+        case 2:
+            // the last 2 symbols (12 bits) are encoding 1 byte (8 bits)
+            // so the last symbol only actually uses 8-6=2 bits and can only take 4 values
+            char last = id.charAt(length - 1);
+            if (last != 'A' && last != 'Q' && last != 'g' && last != 'w') {
+                return false;
+            }
+            break;
+        case 3:
+            // The last 3 symbols (18 bits) are encoding 2 bytes (16 bits)
+            // so the last symbol only actually uses 16-12=4 bits and can only take 16 values
+            last = id.charAt(length - 1);
+            if (last != 'A' && last != 'E' && last != 'I' && last != 'M' && last != 'Q'&& last != 'U'&& last != 'Y'
+                    && last != 'c'&& last != 'g'&& last != 'k' && last != 'o' && last != 's' && last != 'w'
+                    && last != '0' && last != '4' && last != '8') {
+                return false;
+            }
+            break;
+        default:
+            // number & 0x03 is always in [0,3]
+            throw new AssertionError("Impossible case");
+        }
+        for (int i = 0; i < length; ++i) {
+            final char c = id.charAt(i);
+            final boolean allowed =
+                    (c >= '0' && c <= '9') ||
+                    (c >= 'A' && c <= 'Z') ||
+                    (c >= 'a' && c <= 'z') ||
+                    c == '-' || c == '_';
+            if (allowed == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    static boolean isPositiveNumeric(String id) {
+        for (int i = 0; i < id.length(); ++i) {
+            final char c = id.charAt(i);
+            if (c < '0' || c > '9') {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /** With numeric ids, we just fold two consecutive chars in a single byte
+     *  and use 0x0f as an end marker. */
+    private static BytesRef encodeNumericId(String id) {
+        byte[] b = new byte[1 + (id.length() + 1) / 2];
+        b[0] = (byte) NUMERIC;
+        for (int i = 0; i < id.length(); i += 2) {
+            int b1 = id.charAt(i) - '0';
+            int b2;
+            if (i + 1 == id.length()) {
```

Comment:
You could avoid this conditional by checking for an odd length before the loop, and only iterating through the loop on an even number, then have a single condition after the loop to add the last value in the odd case, followed by the end marker.